eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/Relayer.scala

```diff
@@ -222,11 +210,13 @@ object Relayer extends Logging {
   // @formatter:off
   sealed trait ForwardMessage
   case class ForwardAdd(add: UpdateAddHtlc, previousFailures: Seq[AddHtlcFailed] = Seq.empty) extends ForwardMessage
-  case class ForwardFulfill(fulfill: UpdateFulfillHtlc, to: Origin, htlc: UpdateAddHtlc) extends ForwardMessage
-  case class ForwardOnChainFulfill(preimage: ByteVector32, to: Origin, htlc: UpdateAddHtlc) extends ForwardMessage
-  case class ForwardFail(fail: UpdateFailHtlc, to: Origin, htlc: UpdateAddHtlc) extends ForwardMessage
-  case class ForwardFailMalformed(fail: UpdateFailMalformedHtlc, to: Origin, htlc: UpdateAddHtlc) extends ForwardMessage
-  case class ForwardOnChainFail(to: Origin, htlc: UpdateAddHtlc) extends ForwardMessage
+  sealed trait ForwardFulfill extends ForwardMessage { val fulfill: UpdateFulfillHtlc; val to: Origin; val htlc: UpdateAddHtlc }
+  case class ForwardRemoteFulfill(fulfill: UpdateFulfillHtlc, to: Origin, htlc: UpdateAddHtlc) extends ForwardFulfill
+  case class ForwardOnChainFulfill(preimage: ByteVector32, to: Origin, htlc: UpdateAddHtlc) extends ForwardFulfill { override val fulfill  = UpdateFulfillHtlc(htlc.channelId, htlc.id, preimage) }
```

Comment:
why do you generate a fake `fulfill` instead of building a `CMD_FULFILL_HTLC` like in the failure case?

Reply:
Gave it a try, seems to work well (NB: needs to also update `PaymentLifecycle`):

```diff
diff --git a/eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/Relayer.scala b/eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/Relayer.scala
index aedd85664..838bba6a0 100644
--- a/eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/Relayer.scala
+++ b/eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/Relayer.scala
@@ -157,9 +157,9 @@ class Relayer(nodeParams: NodeParams, router: ActorRef, register: ActorRef, comm

     case ff: ForwardFulfill => ff.to match {
       case Origin.Local(_, None) => postRestartCleaner forward ff
-      case Origin.Local(_, Some(sender)) => sender ! ff.fulfill
+      case Origin.Local(_, Some(sender)) => sender ! ff.preimage
       case Origin.Relayed(originChannelId, originHtlcId, amountIn, amountOut) =>
-        val cmd = CMD_FULFILL_HTLC(originHtlcId, ff.fulfill.paymentPreimage, commit = true)
+        val cmd = CMD_FULFILL_HTLC(originHtlcId, ff.preimage, commit = true)
         commandBuffer ! CommandBuffer.CommandSend(originChannelId, cmd)
         context.system.eventStream.publish(ChannelPaymentRelayed(amountIn, amountOut, ff.htlc.paymentHash, originChannelId, ff.htlc.channelId))
       case Origin.TrampolineRelayed(_, None) => postRestartCleaner forward ff
@@ -210,9 +210,9 @@ object Relayer extends Logging {
   // @formatter:off
   sealed trait ForwardMessage
   case class ForwardAdd(add: UpdateAddHtlc, previousFailures: Seq[AddHtlcFailed] = Seq.empty) extends ForwardMessage
-  sealed trait ForwardFulfill extends ForwardMessage { val fulfill: UpdateFulfillHtlc; val to: Origin; val htlc: UpdateAddHtlc }
-  case class ForwardRemoteFulfill(fulfill: UpdateFulfillHtlc, to: Origin, htlc: UpdateAddHtlc) extends ForwardFulfill
-  case class ForwardOnChainFulfill(preimage: ByteVector32, to: Origin, htlc: UpdateAddHtlc) extends ForwardFulfill { override val fulfill  = UpdateFulfillHtlc(htlc.channelId, htlc.id, preimage) }
+  sealed trait ForwardFulfill extends ForwardMessage { val preimage: ByteVector32; val to: Origin; val htlc: UpdateAddHtlc }
+  case class ForwardRemoteFulfill(fulfill: UpdateFulfillHtlc, to: Origin, htlc: UpdateAddHtlc) extends ForwardFulfill { override val preimage = fulfill.paymentPreimage }
+  case class ForwardOnChainFulfill(preimage: ByteVector32, to: Origin, htlc: UpdateAddHtlc) extends ForwardFulfill
   sealed trait ForwardFail extends ForwardMessage { val to: Origin; val htlc: UpdateAddHtlc }
   case class ForwardRemoteFail(fail: UpdateFailHtlc, to: Origin, htlc: UpdateAddHtlc) extends ForwardFail
   case class ForwardRemoteFailMalformed(fail: UpdateFailMalformedHtlc, to: Origin, htlc: UpdateAddHtlc) extends ForwardFail
```

Reply:
Ate my own dogfood and implemented it in e042660da. For even more consistency we could also forward the raw `ForwardFail` to payment senders in the failure case, but that would mean slightly more work in their handling. Trade-off...

Reply:
Heh this was exactly my first implementation, but then I changed it because I preferred keeping the payment lifecycle as untouched as possible and containing the change in the relayer (because pay FSM is already doing a lot, and relayer is simpler).
I had the content of `translateFail` inside a handler for `ForwardFail` in the pay FSM.

If you think that's better I can do that, let me know!

Reply:
I think it is a better because building a fake `UpdateFulfillHtlc` is bad. It's like a trap for future maintainers. The payment lifecycle is only lightly touched, I had a formatting error in my commit, fixed in a566a27ec.

Reply:
Allright let me amend the last commit then.

Reply:
Done in https://github.com/ACINQ/eclair/pull/1349/commits/7d8020c2a3e772e4158bf3c4ba449e282292ddaf

![no need to go deeper](https://user-images.githubusercontent.com/31281497/77155497-e3a6f180-6a9d-11ea-89bd-de42c4786dcd.jpg)